File: problem_3.cpp
// Time Complexity : O(log n)  n=size of window
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Solution {
public:
    int search(const ArrayReader& reader, int target) {
        int l = 0;
        int h = 10;
        
        //need to fix low and high bounds
        
        while (target > reader.get(h)) {
            l = h + 1;
            h = h * 2;
        }
        
        while (l <= h) {
            int mid = l + (h-l)/2;
            
            if(reader.get(mid) == target) return mid;
            
            if (reader.get(mid) > target) {
                //left side
                h = mid -1;
            } else {
                l = mid + 1;
            }
        }
        
        return -1;
    }
};

File: Problem_3.cs
class Solution {

   
    // Time Complexity : O(log m) + O(log n)
    // Space Complexity : O(1)
    // Did this code successfully run on Leetcode : Did not run as i do not have premium but ran through all the test cases.
    // Any problem you faced while coding this : No
    public int search(ArrayReader reader, int target) {
        int low = 0, high = 1;
        //First step to find the range
        while (reader.get(high) < target) {
            low = high;
            high = high * 2;
        }

        //binary search
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (reader.get(mid) == target)
                return mid;
            else if (reader.get(mid) < target)
                high = mid - 1;
            else
                low = mid + 1;
        }
        return -1;
    }
}
